HDDS-3309. Add TimedOutTestsListener to surefire and add timeout to integration tests

smengcl opened the pull request:
## What changes were proposed in this pull request?

Add `TimedOutTestsListener` as a listener to maven-surefire-plugin like Hadoop [does](https://github.com/apache/hadoop/blob/1189af4746919774035f5d64ccb4d2ce21905aaa/hadoop-hdfs-project/hadoop-hdfs/pom.xml#L233-L238). And add timeout to integration tests **that are not marked with `@Ignore`** at the time of the patch.

Also previously discussed under https://github.com/apache/hadoop-ozone/pull/750.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3309

## How was this patch tested?

The patch should make the timed out tests print out the stack trace (if any). Checks itself.

### Bonus

To test if the `TimedOutTestsListener` is working, you can change the timeout of a class to an impossibly small value like 5 sec (for an integration test to initialize). Then run that integration test. Make sure the test isn't ignored or it won't run with `mvn test`.

Note that the timeout must be smaller than `surefire.fork.timeout` shown below otherwise the fork will still be killed with no stack trace or thread dump.

https://github.com/apache/hadoop-ozone/blob/2e9bae2f8636974c5c0e11a4a3e04d9cbfbb2288/pom.xml#L228

For example, in `TestOzoneManagerRestart` I removed the `@Ignore` annotation and set the timeout to `5000` ms. Then ran:
```bash
mvn test -e -Dtest=TestOzoneManagerRestart
```

Maven output:
```bash
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.hadoop.ozone.om.TestOzoneManagerRestart
[ERROR] Tests run: 3, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 15.223 s <<< FAILURE! - in org.apache.hadoop.ozone.om.TestOzoneManagerRestart
[ERROR] testRestartOMWithVolumeOperation(org.apache.hadoop.ozone.om.TestOzoneManagerRestart)  Time elapsed: 5.019 s  <<< ERROR!
java.lang.Exception: test timed out after 5000 milliseconds
	at java.lang.Thread.sleep(Native Method)
	at org.apache.hadoop.test.GenericTestUtils.waitFor(GenericTestUtils.java:218)
...
[ERROR] testRestartOMWithBucketOperation(org.apache.hadoop.ozone.om.TestOzoneManagerRestart)  Time elapsed: 5.004 s  <<< ERROR!
java.lang.Exception: test timed out after 5000 milliseconds
	at java.lang.Thread.sleep(Native Method)
	at org.apache.hadoop.test.GenericTestUtils.waitFor(GenericTestUtils.java:218)
...
[ERROR] testRestartOMWithKeyOperation(org.apache.hadoop.ozone.om.TestOzoneManagerRestart)  Time elapsed: 5.002 s  <<< ERROR!
java.lang.Exception: test timed out after 5000 milliseconds
	at java.lang.Thread.sleep(Native Method)
	at org.apache.hadoop.test.GenericTestUtils.waitFor(GenericTestUtils.java:218)
...
[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   TestOzoneManagerRestart.init:90 »  test timed out after 5000 milliseconds
[ERROR]   TestOzoneManagerRestart.init:90 »  test timed out after 5000 milliseconds
[ERROR]   TestOzoneManagerRestart.init:90 »  test timed out after 5000 milliseconds
[INFO]
[ERROR] Tests run: 3, Failures: 0, Errors: 3, Skipped: 0
```
More importantly, the `surefire-reports` folder has the detailed logs with thread dump:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M1:test (default-test) on project hadoop-ozone-integration-test: There are test failures.
[ERROR]
[ERROR] Please refer to /Users/smeng/repo/ozone/hadoop-ozone/integration-test/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M1:test (default-test) on project hadoop-ozone-integration-test: There are test failures.

Please refer to /Users/smeng/repo/ozone/hadoop-ozone/integration-test/target/surefire-reports for the individual test results.
Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
```

Example thread dump in `surefire-reports/org.apache.hadoop.ozone.om.TestOzoneManagerRestart-output.txt`:
```
====> TEST TIMED OUT. PRINTING THREAD DUMP. <====

Timestamp: 2020-05-22 06:37:20,788

"qtp494169951-212" daemon prio=5 tid=212 timed_waiting
java.lang.Thread.State: TIMED_WAITING
        at sun.misc.Unsafe.park(Native Method)
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
        at org.eclipse.jetty.util.BlockingArrayQueue.poll(BlockingArrayQueue.java:382)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.idleJobPoll(QueuedThreadPool.java:875)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:925)
        at java.lang.Thread.run(Thread.java:748)
"pool-56-thread-1"  prio=5 tid=192 timed_waiting
java.lang.Thread.State: TIMED_WAITING
        at sun.misc.Unsafe.park(Native Method)
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
        at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
        at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
        at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
...
```

smengcl commented:
> Thanks @smengcl for working on this. I tested the change by configuring very short timeout. Here are my observations:
> 
> 1. The `parallel-tests` profile is not enabled by default.  The listener should be configured globally, outside of profiles to take effect.  I think it should be added to the root `pom.xml` here:
> 
> https://github.com/apache/hadoop-ozone/blob/c9a1925537312eb950d596dbae2c7e4d91f9fd59/pom.xml#L1686-L1724
> 
> 1. `TimedOutTestsListener` is not notified of "fork timeouts", only of individual test timeout, ie. the kind which you wanted to add in #750.  So I think we need both changes.

Thanks for the comment @adoroszlai .

I also played with the maven surefire listener a bit. Some observations:

1. `testStarted()` will always be executed at test start, as in its name. But the current `TimedOutTestsListener` doesn't implement it. We can flush the output to make the message appear in e.g. `org.apache.hadoop.ozone.om.TestOzoneManagerRestart-output.txt` even if the test is eventually shutdown by fork timeout (`forkedProcessTimeoutInSeconds`). This way we can print the test name at least. Even if the test fork is terminated by timeout (prints `Killing self fork JVM. Received SHUTDOWN command from Maven shutdown hook.` in `<DATETIME>-jvmRunX.dump`), we can know which test was it running in the output text file.

2. `testFailure` will not be executed when fork JVM for testing timed out. There appears to be one forked JVM **per test class** (See help for [<forkMode>](https://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html). CMIIW).

3. I'm wondering if we could utilize shutdownHook to capture the failed test name and other info. Probably not. But I'll dig a bit.

smengcl commented:
Looks like hadoop-hdds can't locate the class in hadoop-ozone

adoroszlai commented:
> Looks like hadoop-hdds can't locate the class in hadoop-ozone

I started working on a similar listener (for debugging disk space usage in tests) and ran into this classpath issue.  Solved it by moving these test utilities to a new module.  I'll create a PR with the base change to unblock you.  With that change you'll be able to add the `TimedOutTestsListener` globally.

smengcl commented:
> > Looks like hadoop-hdds can't locate the class in hadoop-ozone
> 
> I started working on a similar listener (for debugging disk space usage in tests) and ran into this classpath issue. Solved it by moving these test utilities to a new module. I'll create a PR with the base change to unblock you. With that change you'll be able to add the `TimedOutTestsListener` globally.

It looks to me the root cause of the issue is that `Apache Hadoop HDDS Config` is compiled before `Apache Hadoop Distributed Data Store Common`(`hadoop-hdds-common`). The latter has the `TimedOutTestsListener`. One easy fix might just be to move the `TimedOutTestsListener` class to a module that is compiled early enough.

adoroszlai commented on hadoop-hdds/common/src/test/java/org/apache/hadoop/test/TimedOutTestsListener.java:
```diff
@@ -31,6 +31,7 @@
 
 import org.apache.hadoop.util.StringUtils;
 
+import org.junit.runner.Description;
```
Unused import.

smengcl commented:
@adoroszlai  I have rebased the jira and added commits similar to #750 . I have also left all ignored integration test classes untouched from the timeout change.

Please take a look. :)

smengcl commented:
Rebased again since HDDS-3602 is merged.

smengcl commented:
> Thanks @smengcl for continuing work on this. There are few files with whitespace-only change, and several with ignored test class but timeout being added. Please let me know if you would like a list of these. Otherwise it looks good to me.

Thanks @adoroszlai for another review. I have addressed the whitespace and removed timeout from more ignored tests.

adoroszlai commented:
> I have addressed the whitespace and removed timeout from more ignored tests.

Thanks, the update looks good.

smengcl commented:
Thanks @elek for triggering the recheck. The latest [it-client failure](https://github.com/apache/hadoop-ozone/pull/813/checks?check_run_id=716257941) is caused by flakiness of `TestKeyInputStream#testSeekRandomly` and `TestOzoneRpcClient`.

But I see a green run on my dev branch: https://github.com/smengcl/hadoop-ozone/commit/843f1d894ccfae78a717bd5f5f1faa2cf782d4bc

Will merge shortly.

elek commented:
In case of flaky tests I would suggest disabling them (with opening a Jira + saving log + commit @ignore without review) and repeat test. I think it provides better safety.
